lib/ambient_context.ml

```diff
@@ -18,6 +17,7 @@ let get k =
    Store.get k
 
 
+(* Why is bind neccessary to set a value? *)
```

Comment:
Other TLS's I've seen like the one in `domainslib` has a simple `set` function to add a thread local value, is there a reason why adding the callback is part of the design?

Reply:
This has actually been a lot of the work behind this — Lwt and Eio both expose a very strange (if you don't know the context) 'stack-delineating' API: "during this stack-frame, asynchronous _binds_ will infect their promises with this additional context."

I have my suspicions about the reasons for that design (apparently it was originally Jeremie Dimino who was responsible for it; but the actual plan is lost to history, afaict) … but at the end of the day, one can be converted to the other trivially …

```ocaml
let with_binding k v cb =
   let old_value = get k in
   set k v;
   cb ();
   set k old_value
```

… but the other cannot be converted to the one; and thus a design that allows for both styles of backend is forced on us.

Nice catch, though! I'm surprised this was noticeable upon a casual reading of the code — perhaps disappointed, even.

Reply:
Having thought about this more, I realize that `set` was an api of domain-local storage which has principally different constraints as thread-local storage. Since domains are separate processes with their own isolated memory, they can offer the flexibility of having a direct-style `set` without concern for restoring the previous state.

This is not the case for lightweight-threads because they share the same process memory and therefore the notion of what is "thread-local" is more accurately: _what happened prior in this sequence of binds_. In other words, if you imagine the concurrent program as a directed-acyclic graph with nodes representing branching points. The program can execute and have interleaving branches which can potentially corrupt each other if they retrieve or update the same "thread-local" key. So requiring a callback seems more or less necessary to capture and restore "snapshots" of the thread-local state before executing the next callback